File: instabiz/instabiz/report/ib_pending_approvals/ib_pending_approvals.py
"""IB Pending Approvals — everything waiting on an approver in one place:
journal entries and payments sent for approval, advances on orders, leave,
overtime and expense claims. Journal / payment rows can be approved or
rejected right here; the others open their document."""
import frappe
from frappe import _

from instabiz.overrides.approvals import get_pending


def execute(filters=None):
	f = frappe._dict(filters or {})
	rows = get_pending()
	if f.kind:
		rows = [r for r in rows if r.kind == f.kind]
	if f.only_mine:
		rows = [r for r in rows if r.can_act]
	for r in rows:
		r.action = r.name if r.can_act else ""
	cols = [
		{"label": _("What"), "fieldname": "kind", "fieldtype": "Data", "width": 140},
		{"label": _("Type"), "fieldname": "doctype", "fieldtype": "Data", "hidden": 1},
		{"label": _("Document"), "fieldname": "name", "fieldtype": "Dynamic Link", "options": "doctype", "width": 170},
		{"label": _("Date"), "fieldname": "date", "fieldtype": "Date", "width": 95},
		{"label": _("Party / Employee"), "fieldname": "who", "fieldtype": "Data", "width": 200},
		{"label": _("Amount / Days / Hours"), "fieldname": "amount", "fieldtype": "Float", "width": 130},
		{"label": _("Made By"), "fieldname": "by", "fieldtype": "Link", "options": "User", "width": 170},
		{"label": _("Action"), "fieldname": "action", "fieldtype": "Data", "width": 170},
	]
	kinds = {}
	for r in rows:
		kinds[r.kind] = kinds.get(r.kind, 0) + 1
	summary = [{"label": k, "value": v, "datatype": "Int", "indicator": "orange"} for k, v in kinds.items()] or [
		{"label": _("Waiting"), "value": 0, "datatype": "Int", "indicator": "green"}]
	return cols, rows, None, None, summary


@frappe.whitelist()
def pending_count():
	"""Number card: items the current user can act on."""
	return sum(1 for r in get_pending() if r.can_act)

File: instabiz/instabiz/report/ib_pending_approvals/ib_pending_approvals.js
frappe.query_reports["IB Pending Approvals"] = {
	filters: [
		{ fieldname: "kind", label: __("What"), fieldtype: "Select",
			options: ["", "Accounting entry", "Advance on order", "Leave", "Overtime", "Expense claim"].join("\n") },
		{ fieldname: "only_mine", label: __("Only what I can approve"), fieldtype: "Check", default: 1 },
	],
	formatter(value, row, column, data, default_formatter) {
		if (column.fieldname === "action" && data && data.action) {
			if (["Journal Entry", "Payment Entry"].includes(data.doctype)) {
				const dt = encodeURIComponent(data.doctype), n = encodeURIComponent(data.name);
				return `<a class="btn btn-xs btn-primary" onclick="ib_decide('${dt}','${n}',1)">${__("Approve")}</a>
					<a class="btn btn-xs btn-default" onclick="ib_decide('${dt}','${n}',0)">${__("Reject")}</a>`;
			}
			return `<a class="btn btn-xs btn-default" href="/app/${frappe.router.slug(data.doctype)}/${encodeURIComponent(data.name)}">${__("Open")}</a>`;
		}
		return default_formatter(value, row, column, data);
	},
};

window.ib_decide = function (dt, name, approve) {
	dt = decodeURIComponent(dt);
	name = decodeURIComponent(name);
	const go = (note) => frappe.call({ method: "instabiz.overrides.approvals.decide", args: { doctype: dt, name, approve, note }, freeze: true })
		.then(() => { frappe.show_alert({ message: approve ? __("Approved") : __("Rejected"), indicator: approve ? "green" : "orange" }); frappe.query_report.refresh(); });
	if (approve) frappe.confirm(__("Approve and submit {0}?", [name]), () => go());
	else frappe.prompt({ fieldtype: "Small Text", fieldname: "note", label: __("Reason"), reqd: 1 }, (v) => go(v.note), __("Reject {0}", [name]));
};
